### PR TITLE
[ROCm] Enable unit tests known to work on ROCm

### PR DESCRIPTION
**Pull request by iotamudelta**
* Enable unit tests known to work on ROCm.
* Disable a few that are known to be flaky for the time being.
* Use std::abs for Half
* No more special casing for ROCm in TensorMathReduce
* Document an important detail for a hardcoded block size w.r.t. ROCm in TensorMathReduce

@ezyang @bddppq for awareness

### CODE REVIEW

**Comment by iotamudelta** [comment]
concerning the flake fallout: this is related to a user-contributed patch to make tests work on ROCm with python 3.5.

File: test/test_nn.py [hunk 1]
```diff
@@ -3982,7 +3983,7 @@ def test_NLLLoss_mismatched_batch(self):
         with self.assertRaisesRegex(ValueError, 'Expected.*batch_size'):
             F.nll_loss(x, t)
 
-    @unittest.skipIf(not (TEST_CUDNN and TEST_CUDNN_VERSION >= 7000), "needs cudnn >= 7.0")
+    @unittest.skipIf(not (TEST_CUDNN and (TEST_CUDNN_VERSION if TEST_CUDNN_VERSION else 0) >= 7000), "needs cudnn >= 7.0")
```
Comment (ezyang) [inline test/test_nn.py]: blugh, can we funcify this

**Comment by ezyang** [comment]
yeah, lint is gonna need to be fixed before we can merge

**Review by ezyang — DISMISSED** [review]
lint

File: aten/src/THC/THCNumerics.cuh [hunk 2]
```diff
@@ -205,7 +205,7 @@ struct THCNumerics<at::Half> {
   static inline __host__ __device__ at::Half erf(at::Half a) { return ::erf(a); }
   static inline __host__ __device__ at::Half erfc(at::Half a) { return ::erfc(a); }
   static inline __host__ __device__ at::Half erfinv(at::Half a) { return ::erfinv(a); }
-  static inline __host__ __device__ at::Half abs(at::Half a) { return ::abs(a); }
+  static inline __host__ __device__ at::Half abs(at::Half a) { return std::abs(a); }
```
Comment (bddppq) [inline aten/src/THC/THCNumerics.cuh]: Is this a recent change in the compiler? I remember `std::` was not supported before.

Comment (iotamudelta) [reply aten/src/THC/THCNumerics.cuh]: It's still in general the case that `std::` math functions are not supported and I do not see much indication of that changing any time soon. However, in this particular instance hcc prefers us to use `std::abs()` for the call, i.e., it warns about the `::abs()` usage.

**Comment by ezyang** [comment]
Lint is still broken

**Comment by iotamudelta** [comment]
@ezyang should be fixed now, sorry for that.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**